File: sdk/javascript/src/storage.ts
import type { CloudPool } from './client';
import type { Bucket, FileMetadata, FileSearchResult, UploadResult } from './types';

export class StorageClient {
  constructor(private readonly cp: CloudPool) {}

  /** List all storage buckets for the current user */
  async listBuckets(): Promise<Bucket[]> {
    return this.cp.request<Bucket[]>('GET', '/api/files/buckets');
  }

  /** Create a new storage bucket */
  async createBucket(name: string): Promise<Bucket> {
    return this.cp.request<Bucket>('POST', '/api/files/buckets', { name });
  }

  /**
   * Upload a file to a bucket.
   * Works in both browser (File/Blob) and Node.js (Buffer) environments.
   */
  async upload(bucketName: string, file: File | Blob, fileName?: string): Promise<UploadResult> {
    const formData = new FormData();
    formData.append('file', file, fileName);
    formData.append('bucketName', bucketName);

    const headers: Record<string, string> = {};
    const apiKey = (this.cp as any).apiKey;
    const token = (this.cp as any).authToken;
    if (apiKey) headers['X-API-Key'] = apiKey;
    else if (token) headers['Authorization'] = `Bearer ${token}`;

    const response = await fetch(`${this.cp.getBaseUrl()}/api/files/upload`, {
      method: 'POST',
      headers,
      body: formData,
    });

    if (!response.ok) {
      throw new Error(`Upload failed (${response.status}): ${await response.text()}`);
    }
    return response.json();
  }

  /** List all files in a bucket */
  async listFiles(bucketName: string): Promise<FileMetadata[]> {
    return this.cp.request<FileMetadata[]>('GET', `/api/files?bucketName=${bucketName}`);
  }

  /** Delete a file by ID */
  async deleteFile(fileId: string): Promise<void> {
    await this.cp.request('DELETE', `/api/files/${fileId}`);
  }

  /** Generate a temporary shareable download link */
  async share(fileId: string, expiresInHours = 24): Promise<{ shareUrl: string; expiresAt: string }> {
    return this.cp.request('POST', `/api/files/${fileId}/share`, { expiresInHours });
  }

  /** Semantic search across all uploaded files */
  async search(query: string): Promise<FileSearchResult[]> {
    return this.cp.request<FileSearchResult[]>('GET', `/api/vector/search?query=${encodeURIComponent(query)}`);
  }
}

File: sdk/python/cloudpool/storage.py
from __future__ import annotations

from typing import Any


class StorageClient:
    """Storage APIs."""

    def __init__(self, cp):
        self.cp = cp

    def list_buckets(self) -> list[dict[str, Any]]:
        return self.cp.request(
            "GET",
            "/api/files/buckets",
        )

    def create_bucket(self, name: str) -> dict[str, Any]:
        return self.cp.request(
            "POST",
            "/api/files/buckets",
            json={"name": name},
        )

    def upload(
        self,
        bucket_name: str,
        file_obj,
        file_name: str,
    ) -> dict[str, Any]:

        files = {
            "file": (file_name, file_obj)
        }

        data = {
            "bucketName": bucket_name
        }

        return self.cp.request(
            "POST",
            "/api/files/upload",
            data=data,
            files=files,
        )

    def list_files(
        self,
        bucket_name: str,
    ) -> list[dict[str, Any]]:
        return self.cp.request(
            "GET",
            f"/api/files?bucketName={bucket_name}",
        )

    def delete_file(
        self,
        file_id: str,
    ) -> None:
        self.cp.request(
            "DELETE",
            f"/api/files/{file_id}",
        )

    def share(
        self,
        file_id: str,
        expires_in_hours: int = 24,
    ) -> dict[str, Any]:
        return self.cp.request(
            "POST",
            f"/api/files/{file_id}/share",
            json={
                "expiresInHours": expires_in_hours
            },
        )

    def search(
        self,
        query: str,
    ) -> list[dict[str, Any]]:
        return self.cp.request(
            "GET",
            f"/api/vector/search?query={query}",
        )